### PR TITLE
Add support for composition revisions

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Fixes https://github.com/crossplane/crossplane/issues/1481

This PR contains the implementation of the design proposed in https://github.com/crossplane/crossplane/pull/2437. It depends on https://github.com/crossplane/crossplane-runtime/pull/273.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

[contribution process]: https://git.io/fj2m9

I've tested manually that:

- That Composition works as it did previously without the feature flag.
- Updating revisions manually.
- Updating revisions automatically.

In more detail:

* I confirmed that Composition without revisions works as expected by running Crossplane locally without the new `--enable-alpha-composition-revisions` flag and applying the GCP getting started manifests. I then updated the `Composition` to add a new flag (`spec.forProvider.settings.storageAutoResize`) and observed that the composed `CloudSQLInstance` was updated accordingly soon thereafter.
* I confirmed that Composition with revisions and the `Automatic` update policy work as expected by running Crossplane with the new `--enable-alpha-composition-revisions` flag and applying the GCP getting started  and running the above test. I see that the `CompositePostgreSQLInstance` tracks the most current revision as I toggle the `storageAutoResize` field on the `Composition`, and that those changes propagate to the actual `CloudSQLInstance`.
* I confirmed that Composition with revisions and the `Manual` update policy work as expected by running Crossplane with the new `--enable-alpha-composition-revisions` flag and applying the GCP getting started  and running the above test. I see that the `CompositePostgreSQLInstance` does not track the most current revision as I toggle the `storageAutoResize` field on the `Composition`. When I update the XR's revision manually I see the changes propagate to the actual `CloudSQLInstance`.

### CODE REVIEW

**Comment by negz** [comment]
Some further detail on testing:

* I confirmed that Composition without revisions works as expected by running Crossplane locally without the new `--enable-alpha-composition-revisions` flag and applying the GCP getting started manifests. I then updated the `Composition` to add a new flag (`spec.forProvider.settings.storageAutoResize`) and observed that the composed `CloudSQLInstance` was updated accordingly soon thereafter.

File: internal/controller/apiextensions/claim/configurator.go
Comment (negz) [inline internal/controller/apiextensions/claim/configurator.go:99]: There's no logic change here - I just got stuck on this code for a bit and wanted to try make it a bit more readable.

File: internal/controller/apiextensions/composite/api.go
Comment (negz) [inline internal/controller/apiextensions/composite/api.go:202]: I think it would be useful to mark the current revision as such. I'd like to do that in a future PR though - I'll open an issue to track.

File: internal/controller/apiextensions/composite/composed.go
Comment (negz) [inline internal/controller/apiextensions/composite/composed.go:184]: This refactoring to operate on `[]v1.ComposedTemplate` is spurious. I had originally thought this would help with this PR because both compositions and revisions have an identical `ComposedTemplate` type. I neglected to realise that the `CompositionRevision` type is different though, due to being in a different package.

I decided to leave the refactoring in because these types don't actually operate on the whole `*v1.Composition` - just the templates. Ideally I'd break it out into its own commit but it got a little bit mixed up with the actual `CompositionRevision` implementation so I'd prefer to just leave it as is if no one minds too much. 😬

File: internal/controller/apiextensions/composite/reconciler.go
Comment (negz) [inline internal/controller/apiextensions/composite/reconciler.go:401]: This was removed because if we derive our composition from a revision it won't actually have a UID, version, or name.

File: internal/controller/apiextensions/composite/reconciler.go
Comment (negz) [inline internal/controller/apiextensions/composite/reconciler.go:425]: Another bit of drive-by refactoring. I moved this down from where it was at line 391 because we should validate the Composition _before_ we use it to configure the XR - not after.

**Review by negz — COMMENTED** [review]
I've added a few comments to hopefully make reviewing this giant PR a little easier. I'd suggest reviewing it on a commit by commit basis.

File: internal/controller/apiextensions/composite/api.go
Comment (turkenh) [inline internal/controller/apiextensions/composite/api.go:212]: After reading the code here, I'm wondering if there is any functionality of the `revision` field in the `CompositionRevision` spec other than just showing how many revisions were created (excepts reverts to an older state). 
I was expecting it to be somehow used to find the latest but at the same time, I was aware that wouldn't work in case of reverts. Now, I am wondering if we could drop that `revision` field completely or make it functional (i.e. rely on it to find the latest here) by also increasing it in case of reverts.

File: internal/controller/apiextensions/composition/reconciler.go
Comment (turkenh) [inline internal/controller/apiextensions/composition/reconciler.go:145]: nit: shouldn't the event also include information that this error happened during listing revisions (same for get above)? 

For example:
```suggestion
		r.record.Event(comp, event.Warning(reasonCreateRev, errors.Wrap(err, errListRevs)))
```

**Review by turkenh — APPROVED** [review]
Looking pretty good with just a few non-blocking comments.

Also thank you very much for all the details you've provided in the [proposal](https://github.com/crossplane/crossplane/issues/1481), [design doc](https://github.com/crossplane/crossplane/blob/master/design/one-pager-composition-revisions.md), and code comments which extremely helped to understand the logic and helped a lot for reviewing.

File: apis/apiextensions/v1alpha1/revision_types.go
Comment (negz) [reply apis/apiextensions/v1alpha1/revision_types.go:62]: > Couldn't we find another way other than copying all types and fields from composition_types.go here, like an inline struct?

I think an inline struct would be the way to go once they're all in the same package(s) (e.g. v1beta1 and v1). As it stands with this new type being at v1alpha1, it "felt" a little strange to me for an alpha package to import types from a v1 package, but I could be convinced to do so in the name of simplification. One downside would be that if the v1alpha1 package ever did diverge more from the v1 package we'd need to redo all this copying work again.

> Also wondering, how can we ensure that they will always stay identical?

I don't have a good answer, apart from perhaps adding a comment? It's worth noting that we already have this problem for our v1beta1 and v1 types, which must remain identical.

File: cmd/crossplane/core/core.go
Comment (negz) [reply cmd/crossplane/core/core.go:63]: Yeah, I thought about something like this too. Personally I feel that the UX of distinct flags is better, but I agree that makes them tricky to deprecate. One thing we could do would be to keep the flags around for a few releases even after the feature has been promoted to beta.

**Comment by negz** [comment]
Assigning @muvaf and @hasheddan because I need a maintainer approve (in addition to @turkenh's) before I can merge.

File: cluster/crds/pkg.crossplane.io_controllerconfigs.yaml
Comment (hasheddan) [reply cluster/crds/pkg.crossplane.io_controllerconfigs.yaml:326]: @negz I believe it is from the k8s version bumps (which I think is what you are saying) -- looks strictly additive and only enabled via alpha flag, so I'm not concerned 👍🏻

File: cmd/crossplane/core/core.go
Comment (hasheddan) [reply cmd/crossplane/core/core.go:63]: > One thing we could do would be to keep the flags around for a few releases even after the feature has been promoted to beta.

Just noting here that I believe we _must_ keep them for beta, as that would mean they are "on by default", but can still be turned off. I also like the idea of having distinct flags because a configuration value silently not modifying behavior is negative attribute IMO (i.e. I want to break if I tell a service to behave in a way it does not understand) 👍🏻 

@negz I do wonder if it would make sense to add `default:"false"` to the flag so it will show in the help output (I know typically a bool flag will always be `false` if unspecified, but if we follow the k8s "on by default" strategy this bool flag will actually default to `true` in the future). You can also make the help output group all feature flags together using a group (example: https://github.com/upbound/up/blob/8e5965a42eff0285820247abed47cd91f2d26848/cmd/up/uxp/uxp.go#L40)

File: design/one-pager-composition-revisions.md
Comment (hasheddan) [inline design/one-pager-composition-revisions.md:121]: Was there discussion around this change? I seem to remember someone mentioning it would be nice to be able to easily identify the revision number, but I am in support of this change because revision is in the spec and, more importantly for UX, defined as a printer column 👍🏻

File: internal/controller/apiextensions/composite/api.go
Comment (hasheddan) [inline internal/controller/apiextensions/composite/api.go:205]: @negz there should only ever be one revision that matches `currentHash` correct? If so, can we break out of our range early here?

File: internal/feature/feature_test.go
Comment (hasheddan) [inline internal/feature/feature_test.go:40]: Major Nitpick: aren't we checking here that a flag is _not_ enabled on the `Flags` zero value?

**Review by hasheddan — APPROVED** [review]
Awesome work @negz! I don't think any of my comments are blocking 👍🏻 I would support making the proposed updates to flag, but that could be accomplished in a follow-up if desired.

File: design/one-pager-composition-revisions.md
Comment (negz) [reply design/one-pager-composition-revisions.md:121]: The original discussion was in https://github.com/crossplane/crossplane/pull/2437#discussion_r679028275.

I could have sworn I made this change in a distinct commit with the rationale, but apparently not. I ended up going back to this approach because I realised it was more consistent with what we do for `ProviderRevision` and `ConfigurationRevision`. I like the idea of including the revision in the name, but I prefer consistency and as you say we show the revision pretty prominently as a printer column.

File: internal/controller/apiextensions/composite/api.go
Comment (negz) [reply internal/controller/apiextensions/composite/api.go:205]: Yes - good catch.

File: cmd/crossplane/core/core.go
Comment (negz) [reply cmd/crossplane/core/core.go]: > I do wonder if it would make sense to add default:"false" to the flag so it will show in the help output 

Unfortunately while `default` does appear to actually influence the default value it doesn't change the `--help` output in any way. That said, I do think there's a clean way to handle promoting features to beta. Specifically:

* We don't mention `beta` or `alpha` in the flag name. I've changed this to just `--enable-composition-revisions`.
* We use the `group` as you suggested to make it clear whether a feature is alpha or beta.
* When a feature reaches beta we add `default:"true" negatable:""` to the tags.

This means anyone who is enabling the feature by passing `--enable-composition-revisions` at alpha will not be affected when the feature hits beta. Anyone who wants to disable the feature when it hits beta can pass `--no-enable-composition-revisions`.

**Pull request merged**
